include/jemalloc/internal/arena_structs.h

```diff
@@ -98,10 +98,12 @@ struct arena_s {
 	/*
 	 * The arena is allocated alongside its bins; really this is a
 	 * dynamically sized array determined by the binshard settings.
+	 * Enforcing cacheline-alignment to minimize the number of cachelines
+	 * touched on the hot paths.
 	 */
 	JEMALLOC_WARN_ON_USAGE("Do not use this field directly. "
 	                       "Use `arena_get_bin` instead.")
-	bin_t			 all_bins[0];
+	bin_t			 all_bins[0] JEMALLOC_ALIGNED(CACHELINE);
```

Comment:
Apparently this isn't working for msvc:
https://app.travis-ci.com/github/jemalloc/jemalloc/jobs/612074641

Try putting the align attr at the beginning, i.e.  
```JEMALLOC_ALIGNED(CACHELINE) bin_t all_bins[0];```

Reply:
Just realized this also needs to be updated:

https://github.com/jemalloc/jemalloc/blob/dev/src/arena.c#L1669

Should be updated to:
```
size_t arena_size = ALIGNMENT_CEILING(sizeof(arena_t), CACHELINE) +
    sizeof(bin_t) * nbins_total;
 ... // after base alloc's NULL check
assert((char *)&arena->all_bins[nbins_total -1] + sizeof(bin_t) <= (char *)arena + arena_size));
```

Please also verify that, without the ALIGNMENT_CEILING change, the assert can actually catch this issue.

Reply:
I see. The change in `arena_size` makes sense but the assertion may cause error in CI since we have set `all_bins` warning? Any suggestions?

Reply:
That warning can be suppressed. See the example:
https://github.com/jemalloc/jemalloc/blob/756d4df2fd1b5bde025abed50c9b771376d72c6f/src/arena.c#L1715

Reply:
Updated but use `uintptr_t` for pointer calculation.

Reply:
I did some more sanity checking: the align attr of all_bins[0] properly bumped the size of arena_t to a multiple of cacheline, i.e. padding was added correctly at the end of arena_t, to make sure all_bins[0] always start at the end of arena_t. This means the ALIGNMENT_CEILING wasn't necessary since that was accounted for by sizeof(arena_t). However it doesn't hurt to still add it, in case of different behaviors from different compilers.